pandas/core/arrays/_reshaping.py

```diff
@@ -0,0 +1,236 @@
+"""
+Utilities for implementing 2D compatibility for 1D ExtensionArrays.
+"""
+from functools import wraps
+from typing import Tuple
+
+import numpy as np
+
+from pandas._libs.lib import is_integer
+
+
+def implement_2d(cls):
+    """
+    A decorator to take a 1-dimension-only ExtensionArray subclass and make
+    it support limited 2-dimensional operations.
+    """
+    from pandas.core.arrays import ExtensionArray
+
+    # For backwards-compatibility, if an EA author implemented __len__
+    #  but not size, we use that __len__ method to get an array's size.
+    has_size = cls.size is not ExtensionArray.size
+    has_shape = cls.shape is not ExtensionArray.shape
+    has_len = cls.__len__ is not ExtensionArray.__len__
+
+    if not has_size and has_len:
+        cls.size = property(cls.__len__)
+        cls.__len__ = ExtensionArray.__len__
+
+    elif not has_size and has_shape:
+
+        @property
+        def size(self) -> int:
+            return np.prod(self.shape)
+
+        cls.size = size
+
+    orig_copy = cls.copy
```

Comment:
Should these all be private?

Reply:
were inside a function namespace, why would that matter?

Reply:
My bad, was thinking we were in a class body. The `cls` tripped me up.